src/components/ContentWidgets/ToDoList/ToDoListEditor.tsx

```diff
@@ -0,0 +1,52 @@
+import React, { FormEvent, ChangeEvent } from 'react';
+import { WidgetEditorProps } from '../../ContentMapping/ContentMapping';
+import './ToDoList.css';
+
+export const ToDoListEditor: React.FC<WidgetEditorProps> = ({
+    originalContent, 
+    editedContent,
+    setEditedContentOnChange
+}) => {
+    let list = editedContent.toDoList || [];
+    let task_text = "";
+
+    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        task_text = e.target.value;
+    }
+
+    const addToDo = (e: FormEvent) => {
+        e.preventDefault();
+        let newList = [...list];
+        newList.push({index: newList.length, task: task_text});
+        setEditedContentOnChange("toDoList", newList);
+    }
+
+    const removeToDo = (indexToRemove: number) => {
+        let newList = [...list];
+        newList.splice(indexToRemove, 1);
+        for (let i = indexToRemove; i < newList.length; i++) {
+            newList[i].index = i;
+        }
+        setEditedContentOnChange("toDoList", newList);
+    }
+
+    let items = list.map((item) => {
+        return (
+            <li key={item.index}>{item.task}
+                <button onClick={() => removeToDo(item.index!)}>&times;</button>
```

Comment:
Unclear what the `!` does. Elisa had it too and I'm not entirely sure what it is-- I've never seen this syntax used in javascript before.

Reply:
Since I used `?` in the declaration of the key and value of a to-do list item, I used `!` to say "this expression cannot be null or undefined here, so don't complain about the possibility of it being null or undefined". I can see why I wouldn't want to use `?` now.

https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci

Reply:
o shit this changes everything. I've always been typecasting the item (ie `item.index as number`). I learned something today.

Reply:
also aren't you at work what

Reply:
aren't you at work?? my project manager isn't here and he hasn't responded to my message 🙃